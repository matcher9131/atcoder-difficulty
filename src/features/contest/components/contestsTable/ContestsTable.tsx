import type { ReactElement } from "react";
import clsx from "clsx";

export type ContestsTableProps = {
    readonly contestsTableHeader: ReactElement;
    readonly gridColsClassName: string;
    readonly contestRows: readonly ReactElement[];
};

export const ContestsTable = ({
    contestsTableHeader,
    gridColsClassName,
    contestRows,
}: ContestsTableProps): ReactElement => {
    return (
        <table className={clsx("w-full", "grid", gridColsClassName)}>
            {contestsTableHeader}
            <tbody className="contents">{contestRows}</tbody>
        </table>
    );
};
