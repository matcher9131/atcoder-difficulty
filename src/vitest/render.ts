import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
