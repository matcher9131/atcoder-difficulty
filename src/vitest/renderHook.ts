import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
