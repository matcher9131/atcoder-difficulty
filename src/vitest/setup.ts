import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
