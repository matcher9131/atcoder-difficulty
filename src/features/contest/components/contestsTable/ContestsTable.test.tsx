import { describe, it, expect } from "vitest";
import { screen } from "@testing-library/react";
import { render } from "../../../../vitest/render";
import { ContestsTable } from "./ContestsTable";

// src/features/contest/components/contestsTable/ContestsTable.test.tsx

describe("ContestsTable", () => {
    const mockHeader = (
        <thead data-testid="table-header">
            <tr>
                <th>Header</th>
            </tr>
        </thead>
    );
    const mockRows = [
        <tr key="row1" data-testid="contest-row">
            <td>Row 1</td>
        </tr>,
        <tr key="row2" data-testid="contest-row">
            <td>Row 2</td>
        </tr>,
    ];
    const gridColsClassName = "grid-cols-3";

    it("renders table with correct class names", () => {
        const { container } = render(
            <ContestsTable
                contestsTableHeader={mockHeader}
                gridColsClassName={gridColsClassName}
                contestRows={mockRows}
            />,
        );
        const table = container.querySelector("table");
        expect(table).toBeTruthy();
        expect(table?.className).toContain("w-full");
        expect(table?.className).toContain("grid");
        expect(table?.className).toContain(gridColsClassName);
    });

    it("renders contestsTableHeader", () => {
        render(
            <ContestsTable
                contestsTableHeader={mockHeader}
                gridColsClassName={gridColsClassName}
                contestRows={mockRows}
            />,
        );
        expect(screen.getByTestId("table-header")).toBeInTheDocument();
    });

    it("renders contestRows inside tbody", () => {
        render(
            <ContestsTable
                contestsTableHeader={mockHeader}
                gridColsClassName={gridColsClassName}
                contestRows={mockRows}
            />,
        );
        const rowgroups = screen.getAllByRole("rowgroup");
        // tbody is usually the second rowgroup (after thead)
        const tbody = rowgroups[1];
        expect(tbody).toBeInTheDocument();
        expect(tbody.className).toContain("contents");
        const rows = screen.getAllByTestId("contest-row");
        expect(rows).toHaveLength(2);
        expect(rows[0]).toHaveTextContent("Row 1");
        expect(rows[1]).toHaveTextContent("Row 2");
    });

    it("matches snapshot", () => {
        const { container } = render(
            <ContestsTable
                contestsTableHeader={mockHeader}
                gridColsClassName={gridColsClassName}
                contestRows={mockRows}
            />,
        );
        expect(container).toMatchSnapshot();
    });
});
