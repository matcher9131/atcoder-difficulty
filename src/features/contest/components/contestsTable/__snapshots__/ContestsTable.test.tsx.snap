// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContestsTable > matches snapshot 1`] = `
<div>
  <table
    class="w-full grid grid-cols-3"
  >
    <thead
      data-testid="table-header"
    >
      <tr>
        <th>
          Header
        </th>
      </tr>
    </thead>
    <tbody
      class="contents"
    >
      <tr
        data-testid="contest-row"
      >
        <td>
          Row 1
        </td>
      </tr>
      <tr
        data-testid="contest-row"
      >
        <td>
          Row 2
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
